gnssanalysis/gn_download.py

```diff
@@ -211,6 +212,294 @@ def gen_prod_filename(dt, pref, suff, f_type, wkly_file=False, repro3=False):
     return f, gpswk
 
 
+def generate_uncompressed_filename(filename: str) -> str:
+    """Returns a string of the uncompressed filename given the [assumed compressed] filename
+
+    :param str filename: Original filename of compressed file
+    :return str: The uncompressed filename based on input (returns input filename if compression not recognised)
+    """
+    if filename.endswith(".tar.gz") or filename.endswith(".tar"):
+        with _tarfile.open(filename, "r") as tar:
+            # Get name of file inside tar.gz file (assuming only one file)
+            return tar.getmembers()[0].name
+    elif filename.endswith(".crx.gz"):
+        return filename[:-6] + "rnx"
+    elif filename.endswith(".gz"):
+        return filename[:-3]
+    elif filename.endswith(".Z"):
+        return filename[:-2]
+    elif filename.endswith(".bz2"):
+        return filename[:-4]
+    else:
+        logging.debug(f"{filename} not compressed - extension not a recognized compression format")
+        return filename
+
```

Comment:
Running this through testing I realised that there's a bigger problem with how this has been written up: the `tarfile.open` function only works when there is an actual file to open. The `generate_uncompressed_filename` function is supposed to work on strings so this will fail unless the `filename` variable actually points at an existing file. We may have to make an exception for `tar.gz` files in that we cannot generate its uncompressed filename

Reply:
Having no `.` in the filename seems... very unlikely? I shouldn't need to handle e.g. a file named `bz2`.

Reply:
It's probably not realistic to reliably predict the filename of a compressed directory/collection - only the case where a single file is compressed. And even then, there is no universal convention for naming a compressed directory vs a compressed single file. So when provided with only a filename, we'd still have to make assumptions based on what is typical of specific sources.

Reply:
Implemented your first suggestion